nixos/modules/services/monitoring/certspotter.nix

```diff
@@ -0,0 +1,149 @@
+      path = pkgs.writeText "certspotter-email_recipients" (builtins.concatStringsSep "\n" cfg.emailRecipients);
+    }
+    ++ lib.optional (cfg.hooks != [ ]) {
+      name = "hooks.d";
+      path = pkgs.linkFarm "certspotter-hooks" (lib.imap1 (i: path: {
+        inherit path;
+        name = "hook${toString i}";
+      }) cfg.hooks);
+    });
+in
+{
+  options.services.certspotter = {
+    enable = lib.mkEnableOption "Cert Spotter, a Certificate Transparency log monitor";
+
+    package = lib.mkPackageOptionMD pkgs "certspotter" { };
+
+    startAtEnd = lib.mkOption {
+      type = lib.types.bool;
+      description = ''
+        Whether to skip certificates issued before the first launch of Cert Spotter.
+        Setting this to `false` will cause Cert Spotter to download tens of terabytes of data.
+      '';
+      default = true;
+    };
+
+    sendmailPath = lib.mkOption {
+      type = lib.types.path;
+      description = ''
+        Path to the `sendmail` binary. By default, the local sendmail wrapper is used
+        (see {option}`services.mail.sendmailSetuidWrapper`}).
+      '';
+      example = lib.literalExpression ''"''${pkgs.system-sendmail}/bin/sendmail"'';
+    };
+
+    watchlist = lib.mkOption {
+      type = with lib.types; listOf str;
+      description = "Domain names to watch. To monitor a domain with all subdomains, prefix its name with `.` (e.g. `.example.org`).";
+      default = [ ];
+      example = [ ".example.org" "another.example.com" ];
+    };
+
+    emailRecipients = lib.mkOption {
+      type = with lib.types; listOf str;
+      description = "A list of email addresses to send certificate updates to.";
+      default = [ ];
+    };
+
+    hooks = lib.mkOption {
+      type = with lib.types; listOf path;
+      description = ''
+        Scripts to run upon the detection of a new certificate. See `man 8 certspotter-script` or
+        [the GitHub page](https://github.com/SSLMate/certspotter/blob/${pkgs.certspotter.src.rev or "master"}/man/certspotter-script.md)
+        for more info.
+      '';
+      default = [ ];
+      example = lib.literalExpression ''
+        [
+          (pkgs.writeShellScript "certspotter-hook" '''
+            echo "Event summary: $SUMMARY."
+          ''')
+        ]
+      '';
+    };
+
+    extraFlags = lib.mkOption {
+      type = with lib.types; listOf str;
+      description = "Extra command-line arguments to pass to Cert Spotter";
+      example = [ "-no_save" ];
+      default = [ ];
+    };
+  };
+
+  config = lib.mkIf cfg.enable {
+    assertions = [
+      {
+        assertion = cfg.hooks != [ ] || cfg.emailRecipients != [ ]
+          || builtins.any (x: builtins.elem x cfg.extraFlags) [ "-stdout" "-email" "-watchlist" ];
+        message = "You must specify at least one hook or email recipient for Cert Spotter";
+      }
+      {
+        assertion = (cfg.emailRecipients != [ ]) -> (cfg.sendmailPath != "/run/current-system/sw/bin/false");
```

Comment:
Actually, this is not such a nonsensical configuration either: certspotter drops all the found certs in $STATE_DIRECTORY/certs, so one might not set any hooks or notification mechanisms and have something else monitor that directory. (I actually plan to do that, because that seems to be the simplest way of subtracting the expected certs, if the expected certs might arrive with a delay.)

Reply:
in that case, pass the `-stdout` flag instead of setting up hooks or email notifications, or add an empty hook

Reply:
ok, I made it so `sendmailPath` is now null if no wrapper is configured, and you can set it to `/run/current-system/sw/bin/false` manually

Reply:
I don't understand why specifying `-watchlist` via extraFlags satisfies this assertion.

I would prefer if you didn't try parsing extraFlags, because the Golang flag parsing is [not _that_ straighforward](https://pkg.go.dev/flag#hdr-Command_line_flag_syntax). Frankly, I think it's better to keep the old assertion and get rid of it entirely if and when I convince certspotter people that running without notification methods should be fine (which btw they don't actually check for correctly, because running with an empty email file will satisfy [that check](https://github.com/SSLMate/certspotter/blob/master/cmd/certspotter/main.go#L218C31-L218C47)).

Note also that overriding assertions in modules you use is extremely annoying (they go into a list, as opposed to a dict, so one can't just override them by name), so one should use them only for cases where not only we can't think of a reason to use the module in a way that violates the proposed assertion, but where we can convince ourselves under some reasonable standard of proof that it's ~impossible _and_ where they serve a purpose (e.g. preventing a hard-to-notice misconfiguration, or preventing us from building a system image that will not boot). In this case the misconfiguration will be immediately obvious (the service fails to start with an explanatory message) and will not brick the machine.

Reply:
I removed the assertion and made `hooks.d` always generate if no emails are added